lib/rails-mermaid_erd_markdown/markdown.rb

```diff
@@ -0,0 +1,105 @@
+# frozen_string_literal: true
+
+module MermaidErdMarkdown
+  class Markdown
+    attr_accessor :is_show_key, :is_show_comment
+
+    def self.generate(&block)
+      new.generate(&block)
+    end
+
+
+    def initialize
+      @is_show_key = true
+      @is_show_comment = true
+    end
+
+    def add(element)
+      document << element
+    end
+
+    def document
+      @document ||= []
+    end
+
+    def erd(&block)
+      add(erd_header)
+      block.arity == 1 ? block[self] : instance_eval(&block)
+      add(erd_footer)
+    end
+
+    def erd_footer
+      "```"
+    end
+
+    def erd_header
+      [
+        "```mermaid",
+        "erDiagram",
+        "    %% --------------------------------------------------------",
+        "    %% Entity-Relationship Diagram",
+        "    %% --------------------------------------------------------",
+        ""
+      ].join("\n")
+    end
+
+    def erd_relation(relation)
+      left_model_name = relation[:LeftModelName].tr(":", "-")
+      right_model_name = relation[:RightModelName].tr(":", "-")
+      comment = is_show_comment ? ": \"#{relation[:Comment]}\"" : ": \"\""
+
+      "    #{left_model_name} #{relation[:LeftValue]}"\
+      "#{relation[:Line]}#{relation[:RightValue]}"\
+      " #{right_model_name} #{comment}"
+    end
+
+    def erd_table(table_name, model_name)
+      lines = [erd_table_header(table_name, model_name)]
+      lines << yield
+      lines << erd_table_footer
+      lines.join("\n")
+    end
+
+    def erd_table_column(column)
+      key = is_show_key ? column[:key] : ""
+      line = "        #{column[:type]} #{column[:name]}"
+      line << " #{key}" unless key.empty?
+      line
+    end
+
+    def erd_table_header(table_name, model_name)
+      [
+        "    %% table name: #{table_name}",
+        "    #{model_name.tr(":", "-")}{"
+      ].join("\n")
+    end
+
+    def erd_table_footer
+      [
+        "    }",
+        ""
+      ].join("\n")
+    end
+
+    def header(text)
+      "# #{text}\n"
+    end
+
+    def generate(&block)
+      block.arity == 1 ? block[self] : instance_eval(&block)
```

Comment:
A bit confused about what's happening here

Reply:
First it checks if the passed in block is expecting an attribute, then yields to the block in one of two ways.
- With no expected attribute it will process the block as if it is running within the `Markdown` class able to make calls like directly to methods within via `insatnace_eval(&block)`, but would not have access to the calling classes methods like:
```ruby
def body
  "Description"
end

def new_markdown
  MermaidErdMarkdown::Markdown.generate do
    add header("Title") #This would work as it is in scope in the `Markdown` class.
    add body #This would fail as body isn't in scope here.
  end
end
```
- With an attribute passed then I have access to both since `self` is the thing the attribute is being set to. Then you reference the `Markdown` class with the attribute:
```ruby
def body
  "Description"
end

def new_markdown
  MermaidErdMarkdown::Markdown.generate do |doc|
    doc.add doc.header("Title")
    doc.add body #This now works.
  end
end
```

I started with the former first, then changed things to the latter as I needed. Looking back I probably could convert the `mermaid_markdown`'s block to not used the attribute, but that might be more confusing. I could clean this up so it expects an arity of 1 and get rid of the other.

Reply:
How does `block[self]` have access to the surrounding class context? Doesn't `self` here also refer to the `Markdown` instance, since it is called within `Markdown`?

Reply:
It is passing self back to the calling method as the param used there.  So in the example above the `Markdown` class becomes the `doc` param. This encapsulation results in the rest of the block running in the calling instance, thus having access to the rest of the calling instance. And `doc` having access to the `Markdown` instance.